CONTRIBUTING.md

```diff
@@ -0,0 +1,142 @@
+
+
+Note that while we plan to enforce these rules as they are added, a lot of code was produced before they existed and we don't normally reprocess all of it, but opportunistically update it to meet the rules as we update it for other reasons.
+
+Exceptions can be made for `module_utils/` as that code targets older versions of Python than the rest of core and not all rules might not be 100% compatible.
+
+Documentation from docs.ansible.com is mostly designed for collection or 3rd party plugin developers, this document covers core code (this repo) but it also includes guidance on plugin development. If you find any contradictions, keep in mind that most of this applies to core >= 2.19, so which documnent you favor depends on the versions you plan to support.
+
+## PUBLIC API
+
+Anything under `_internal` should not be used by plugins or 3rd party applications, it is subject to change w/o notice and only directly core in code may depend on it.
+
+
+# Error handling
+
+How to handle errors in Ansible.
+
+## Tracebacks
+
+Do not manually generate tracebacks for errors or warnings. Standardized traceback capture is included for errors, warnings and deprecation warnings for both controller and module code (Python). See the `DISPLAY_TRACEBACK` config option for details on how to enable traceback capture and display.
+
+## In modules
+
+In most cases, just raise an exception. The AnsiballZ wrapper now provides a general exception handler for Python modules, making use of the `AnsibleModule.fail_json` method unnecessary, unless the module result needs to be customized. Calls to the various Python-module-side `warn` and `deprecate` methods/functions will also capture and marshal tracebacks to the controller when enabled.
+
+## Deferred exceptions in modules
+
+When deferring exceptions in modules using `try/except` and `fail_json`, pass the captured `Exception` instance to `fail_json` using the `exception` argument. The error handling infrastructure will handle collection of error details and traceback formatting.
+
+## Exception context
+
+Raising exceptions while another exception is active will cause the active exception to become the `__context__` for the newly raised exception. This is usually not the intended behavior, and should be reserved for unexpected errors while handling the original exception. In most cases a `raise from` is desired, either with `None` to suppress the original exception if it's not helpful, or with the captured exception to set it as the `__cause__` for the newly raised exception.
+
+In earlier versions of Ansible, this could be done by setting `orig_exc` on `AnsibleError`. However, this is no longer needed and a simple `raise from` should be used instead.
+
+Suppression example: `raise Exception("something") from None`
+Cause example: `raise Exception("something") from ex`
```

Comment:
```suggestion
In Ansible versions before 2.19, this could be done by setting `orig_exc` on `AnsibleError`. However, this is no longer needed, and a simple `raise from` should be used instead.

Suppression example: `raise Exception("something") from None`
Cause example: `raise Exception("something") from ex`
```